### Importing merged Dataset  

Following our mentor’s recommendation, we separated the data merge step from the cleaning and visualization steps. This makes the workflow more modular and easier to maintain.  

We first download the prepared merge file (covering the years 2019–2023) from Google Drive using `gdown`. Afterwards, we load the dataset into a Pandas DataFrame. While importing, we explicitly define some columns (`nbv`, `pr`, `lartpc`, `larrout`, `id_usager`) as strings, since they contained mixed data types across different years. This ensures consistency and avoids parsing errors later during analysis.  

Next, we clean and reformat specific columns:  
- **`id_vehicule`**: Converted from string to `int64` after removing spaces (`" "`) and non-breaking spaces (`"\xa0"`). This step ensures all vehicle IDs are treated as integers instead of mixed types.  
- **Categorical columns (`grav`, `catu`, `sexe`, `trajet`, `secu1`, `secu2`, `secu3`)**: Originally stored as `float64`, these columns contained missing values. To avoid issues, missing values were replaced with `-1` (representing *not specified*) and the columns were cast to `int64`.  

This preprocessing step results in a more consistent dataset where identifiers are properly formatted, and categorical variables no longer contain floats or missing values.  


In [ ]:
!pip install -U gdown

In [ ]:
import gdown

# ~~~ FILE LOCATIONS ~~~

file_id = '193qtxJgZKQt7QRLqKLQlZUxWOsomgnVA'             # road_accidents_france_2019_2023.csv
url = f'https://drive.google.com/uc?id={file_id}'

# ~~~ DOWNLOAD FILE ~~~

gdown.download(url, 'road_accidents_france_2019_2023.csv', quiet=False)

# ~~~ READ CVS FILE ~~~

import pandas as pd

# Full data
full_data = pd.read_csv('road_accidents_france_2019_2023.csv', sep=';',
                        dtype= {
                            "nbv": "string",
                            "pr": "string",
                            "lartpc": "string",
                            "larrout": "string",
                            "id_usager":"string",
                            }                         # Dealing with Columns (20,23,26,27,54) have mixed types
                        )

# Convert id_vehicule from string to int64 by removing " " and "\xa0"
full_data["id_vehicule"] = (
    full_data["id_vehicule"]
    .astype(str)                                  # Convert id_vehicule from an object to a string
    .str.replace(" ", "", regex=True)             # Example:"138 306 520" -> to convert to int64 we need to remove the space " "
    .str.replace("\xa0", "", regex=False)         # Example:"154\xa0658\xa0350" -> To convert to int64 we need to remove "\xa0"
    .astype("int64")                              # Convert from string to int64
)

# Convert "grav", "catu", "sexe", "trajet", "secu1", "secu2", "secu3" from float64 to int64 by replacing NA's with -1 (Not-specified)
full_data["grav"] = full_data["grav"].fillna(-1).astype('int64')
full_data["catu"] = full_data["catu"].fillna(-1).astype('int64')
full_data["trajet"] = full_data["trajet"].fillna(-1).astype('int64')
full_data["sexe"] = full_data["sexe"].fillna(-1).astype('int64')
full_data["secu1"] = full_data["secu1"].fillna(-1).astype('int64')
full_data["secu2"] = full_data["secu2"].fillna(-1).astype('int64')
full_data["secu3"] = full_data["secu3"].fillna(-1).astype('int64')

Downloading...
From (original): https://drive.google.com/uc?id=193qtxJgZKQt7QRLqKLQlZUxWOsomgnVA
From (redirected): https://drive.google.com/uc?id=193qtxJgZKQt7QRLqKLQlZUxWOsomgnVA&confirm=t&uuid=ef1a3d6f-05f6-4988-adce-c9bdd8f03555
To: /content/road_accidents_france_2019_2023.csv
100%|██████████| 132M/132M [00:02<00:00, 48.8MB/s]
/tmp/ipython-input-657516609.py:17: DtypeWarning: Columns (40,42,44,45,48,51,52,53) have mixed types. Specify dtype option on import or set low_memory=False.
  full_data = pd.read_csv('road_accidents_france_2019_2023.csv', sep=';',


In [ ]:
# Overview Full data
display(full_data.head())

# Data types and missing values
display(full_data.info())

# Count the number of unique accidents recorded
unique_count = full_data["Num_Acc"].nunique()
print(f"The number of accidents recorded in the full data is: {unique_count}")

Num_Acc  id_vehicule num_veh  place  catu  grav  sexe  an_nais  \
0  201900000001    138306524     B01      2     2     4     2   2002.0   
1  201900000001    138306524     B01      1     1     4     2   1993.0   
2  201900000001    138306525     A01      1     1     1     1   1959.0   
3  201900000002    138306523     A01      1     1     4     2   1994.0   
4  201900000003    138306520     A01      1     1     1     1   1996.0   

   trajet  secu1  ...  prof  pr  pr1 plan  lartpc larrout  surf  infra  situ  \
0       0      1  ...     1   6  900    2    <NA>    <NA>     1      2     1   
1       5      1  ...     1   6  900    2    <NA>    <NA>     1      2     1   
2       0      1  ...     1   6  900    2    <NA>    <NA>     1      2     1   
3       0      1  ...     4   3  845    2    <NA>    <NA>     1      0     1   
4       0      1  ...     1  10  500    3    <NA>    <NA>     1      0     1   

   vma  
0   70  
1   70  
2   70  
3   70  
4   90  

[5 rows x 55 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 619807 entries, 0 to 619806
Data columns (total 55 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   Num_Acc      619807 non-null  int64  
 1   id_vehicule  619807 non-null  int64  
 2   num_veh      619807 non-null  object 
 3   place        619807 non-null  int64  
 4   catu         619807 non-null  int64  
 5   grav         619807 non-null  int64  
 6   sexe         619807 non-null  int64  
 7   an_nais      611268 non-null  float64
 8   trajet       619807 non-null  int64  
 9   secu1        619807 non-null  int64  
 10  secu2        619807 non-null  int64  
 11  secu3        619807 non-null  int64  
 12  locp         619807 non-null  int64  
 13  actp         619807 non-null  object 
 14  etatp        619807 non-null  int64  
 15  id_usager    381699 non-null  string 
 16  senc         619807 non-null  int64  
 17  catv         619807 non-null  int64  
 18  obs          619807 non-

None

The number of accidents recorded in the full data is: 273226


### Helper Function: Checking Unique Values  

To support our data exploration and cleaning process, we defined a helper function `checkColumn()`.  
This function returns the **counts** and **percentages** of unique values in a given column, including missing values.  

It also offers two sorting options:  
- `sort=0` → sort results by the column index (ascending)  
- `sort=1` → sort results by frequency (descending, default)  

This allows us to quickly inspect the distribution of values across different variables and identify potential anomalies or imbalances in the dataset.  


In [ ]:
# Function to check unqiue values and their percentages for a certain column
def checkColumn(df, col, sort=1):
    """
    Show counts + percentages for a column.

    Parameters
    ----------
    df : pandas.DataFrame
        The dataframe
    col : str
        Column name
    sort : int, optional
        0 → sort by index (ascending)
        1 → sort by count (descending, default)
    """
    counts = df[col].value_counts(dropna=False)
    percentages = df[col].value_counts(normalize=True, dropna=False).mul(100).round(2)

    result = pd.DataFrame({'count': counts, 'percentage': percentages})

    if sort == 0:
        result = result.sort_index()
    elif sort == 1:
        result = result.sort_values("count", ascending=False)

    return result

# Example
display(checkColumn(full_data, 'sexe'))


count  percentage
sexe                    
 1    417805       67.41
 2    193765       31.26
-1      8237        1.33

### Dropping, Renaming, Retyping columns
After analyzing the different tables in detail and discussing with our project mentor, we concluded that not all columns were essential for our project objectives. Therefore, we decided it was best to remove some of them.

The reasons for removing a column were:

The data was irrelevant to our priority focus.

The data was too specific (e.g. voie, which refers to the exact road name where the accident occurred).

The column had too much missing data (e.g. v2 has 91% missing values).

The following columns were dropped:

Characteristics: adr

Vehicles: senc, occutc

Users: place, locp, actp, etatp, id_usager

Places: voie, v1, v2, pr, pr1, lartpc, larrout

Since the initial cvs files are in French, it is essential to translate all column names into English so that we can easily understand and work with them.

 We first agreed on a consistent naming style: for example, whenever a column refers to a “number” we place it at the beginning using the prefix num_ (e.g. the original column nbv became num_traffic_lane). These decisions were defined and documented in our data audit.

Even though our dataset at this stage contains 40 columns, we structured the renaming code by grouping the columns under their original table categories (Characteristics, Vehicles, Users, and Places). This way, any group member can quickly return to this section of the notebook, see where each column came from, and easily make modifications if needed.

In [ ]:
# Create a new dataframe called 'data' and drop the columns that are mentioned above
data = full_data.drop(columns = ['adr','senc','occutc','place','locp','actp','etatp','id_usager','voie','v1','v2','pr','pr1','lartpc','larrout'])



# Rename the columns that are written in French with the new English names found in our Data Audit
data = data.rename(columns = {
    "Num_Acc": "num_accident",      # Renaming columns initially from Characteristics Table
    "jour": "day",
    "mois": "month",
    "lum": "light_cond",
    "agg": "loc_type",
    "int": "intersc_type",
    "atm": "weather_cond",
    "col": "collision_type"
})

data = data.rename(columns = {
    "senc": "direc_travel",         # Renaming columns initially from Vehicles Table
    "catv": "category_vehicle",
    "obs": "fixed_obs_hit",
    "obsm": "mobile_obs_hit",
    "choc": "int_impact_point",
    "manv": "maneuvre"
})

data = data.rename(columns = {
    "id_vehicule": "id_vehicle",    # Renaming columns initially from Users Table
    "num_veh": "num_vehicle",
    "catu": "category_user",
    "grav": "injury_severity",
    "sexe": "gender",
    "an_nais": "year_of_birth",
    "trajet": "trip_purpose",
    "secu1": "safety_equipment1",
    "secu2": "safety_equipment2",
    "secu3": "safety_equipment3"
})

data = data.rename(columns = {
    "catr": "category_road",      # Renaming columns initially from Places Table
    "circ": "traffic_reg",
    "nbv": "num_traffic_lane",
    "vosp": "reserved_lane",
    "prof": "road_slope",
    "plan": "road_alignmt",
    "surf": "road_surface",
    "situ": "accident_situ",
    "vma": "speed_limit"
})

# Convert the column 'hrmn' to a time format
data['hrmn'] = pd.to_datetime(data["hrmn"], format = "%H:%M", errors="coerce").dt.time
print("Verification that hrmn is now in time format:" , "\n" , data["hrmn"].head())

# Verify that new dataframe called 'data' should have 40 columns and new column names
display(data.info())

Verification that hrmn is now in time format: 
 0    01:30:00
1    01:30:00
2    01:30:00
3    02:50:00
4    15:15:00
Name: hrmn, dtype: object
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 619807 entries, 0 to 619806
Data columns (total 40 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   num_accident       619807 non-null  int64  
 1   id_vehicle         619807 non-null  int64  
 2   num_vehicle        619807 non-null  object 
 3   category_user      619807 non-null  int64  
 4   injury_severity    619807 non-null  int64  
 5   gender             619807 non-null  int64  
 6   year_of_birth      611268 non-null  float64
 7   trip_purpose       619807 non-null  int64  
 8   safety_equipment1  619807 non-null  int64  
 9   safety_equipment2  619807 non-null  int64  
 10  safety_equipment3  619807 non-null  int64  
 11  category_vehicle   619807 non-null  int64  
 12  fixed_obs_hit      619807 non-null  int64  
 13  mobil

None

In [ ]:
data.head()

num_accident  id_vehicle num_vehicle  category_user  injury_severity  \
0  201900000001   138306524         B01              2                4   
1  201900000001   138306524         B01              1                4   
2  201900000001   138306525         A01              1                1   
3  201900000002   138306523         A01              1                4   
4  201900000003   138306520         A01              1                1   

   gender  year_of_birth  trip_purpose  safety_equipment1  safety_equipment2  \
0       2         2002.0             0                  1                  0   
1       2         1993.0             5                  1                  0   
2       1         1959.0             0                  1                  0   
3       2         1994.0             0                  1                  0   
4       1         1996.0             0                  1                  0   

   ...  category_road  traffic_reg  num_traffic_lane  reserved_lane  \
0  ...              1            3                10              0   
1  ...              1            3                10              0   
2  ...              1            3                10              0   
3  ...              1            1                 2              0   
4  ...              1            3                 8              0   

   road_slope  road_alignmt  road_surface  infra  accident_situ speed_limit  
0           1             2             1      2              1          70  
1           1             2             1      2              1          70  
2           1             2             1      2              1          70  
3           4             2             1      0              1          70  
4           1             3             1      0              1          90  

[5 rows x 40 columns]

### Creating Age and Age Buckets  

To make age-related analysis easier, we first calculated the **age** of each user by subtracting their year of birth from the accident year. We ensured data quality by coercing invalid entries to `NaN` and storing the result as an integer type (`Int64`), which also supports missing values.  

Since plotting every single age would create too much clutter in visualizations, we grouped ages into **age buckets**. These buckets represent broader ranges (e.g., *18–24*, *25–34*, *85 and older*) and provide a clearer overview of demographic patterns in the dataset.  

Finally, we used the `checkColumn()` function to verify the distribution across buckets and confirm that the transformation was applied correctly.  


In [ ]:
# Import numpy
import numpy as np

# Add age
data['age'] = data['year'] - data['year_of_birth']

# Force only valid numbers
data['age'] = pd.to_numeric(data['age'], errors='coerce')

# Convert to integer
data['age'] = data['age'].astype("Int64")

# Add age_bucket
labels = ['0–13','14–17','18–24','25–34','35–44', '45–54','55–64','65–74','75–84','85 and older']
bins = [0, 13, 17, 24, 34, 44, 54, 64, 74, 84, np.inf]
data['age_bucket'] = pd.cut(data['age'], bins=bins, right=True, labels=labels, include_lowest=True)

# Display age and age_bucket
display(data[['num_accident', 'year', 'year_of_birth', 'age', 'age_bucket']].head())

#Check columns
display(checkColumn(data, 'age_bucket', sort=1))
display(data.info())


num_accident  year  year_of_birth  age age_bucket
0  201900000001  2019         2002.0   17      14–17
1  201900000001  2019         1993.0   26      25–34
2  201900000001  2019         1959.0   60      55–64
3  201900000002  2019         1994.0   25      25–34
4  201900000003  2019         1996.0   23      18–24

count  percentage
age_bucket                      
25–34         127825       20.62
18–24         115979       18.71
35–44          97596       15.75
45–54          84939       13.70
55–64          63393       10.23
65–74          36479        5.89
0–13           28654        4.62
14–17          26798        4.32
75–84          21182        3.42
NaN             8539        1.38
85 and older    8423        1.36

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 619807 entries, 0 to 619806
Data columns (total 42 columns):
 #   Column             Non-Null Count   Dtype   
---  ------             --------------   -----   
 0   num_accident       619807 non-null  int64   
 1   id_vehicle         619807 non-null  int64   
 2   num_vehicle        619807 non-null  object  
 3   category_user      619807 non-null  int64   
 4   injury_severity    619807 non-null  int64   
 5   gender             619807 non-null  int64   
 6   year_of_birth      611268 non-null  float64 
 7   trip_purpose       619807 non-null  int64   
 8   safety_equipment1  619807 non-null  int64   
 9   safety_equipment2  619807 non-null  int64   
 10  safety_equipment3  619807 non-null  int64   
 11  category_vehicle   619807 non-null  int64   
 12  fixed_obs_hit      619807 non-null  int64   
 13  mobile_obs_hit     619807 non-null  int64   
 14  int_impact_point   619807 non-null  int64   
 15  maneuvre           619807 non-null

None

### Creating Date and Time Columns  

The dataset originally provided separate fields for **day**, **month**, **year**, and **time**.  
To perform meaningful temporal analysis, we combined these into new columns:  

- **`date`** → combines day, month, and year into a standard date format.  
- **`datetime`** → extends `date` by also including the `hrmn` field (hour and minute).  
- **`hour`** → extracts the hour of the day from `datetime`, allowing us to analyze accident frequency by time of day.  
- **`weekday`** → derives the day of the week from `date`, stored as a categorical variable for efficient grouping and clear visualization.  

These transformations allow us to explore patterns such as whether accidents are more frequent on weekends, during specific weekdays, or at certain hours of the day.  


In [ ]:
#Add date
data['date'] = pd.to_datetime(
    data['year'].astype(str) + '-' + data['month'].astype(str) + '-' + data['day'].astype(str))

#Add datetime
data['datetime'] = pd.to_datetime(
    data['year'].astype(str) + '-' + data['month'].astype(str) + '-' + data['day'].astype(str) + '-' + data['hrmn'].astype(str))

#Add hour of the day
data['hour'] = data['datetime'].dt.hour

#Add day of week
data['weekday'] = data['date'].dt.day_name().astype('category')

#Display date, datetime and hour of the day
display(data[['num_accident', 'day', 'month', 'hrmn', 'year','date', 'datetime', 'hour', 'weekday']].head())

#Check columns
display(checkColumn(data, 'hour', sort=1))
display(checkColumn(data, 'weekday', sort=1))
display(data.info())


num_accident  day  month      hrmn  year       date            datetime  \
0  201900000001   30     11  01:30:00  2019 2019-11-30 2019-11-30 01:30:00   
1  201900000001   30     11  01:30:00  2019 2019-11-30 2019-11-30 01:30:00   
2  201900000001   30     11  01:30:00  2019 2019-11-30 2019-11-30 01:30:00   
3  201900000002   30     11  02:50:00  2019 2019-11-30 2019-11-30 02:50:00   
4  201900000003   28     11  15:15:00  2019 2019-11-28 2019-11-28 15:15:00   

   hour   weekday  
0     1  Saturday  
1     1  Saturday  
2     1  Saturday  
3     2  Saturday  
4    15  Thursday

count  percentage
hour                   
17    55459        8.95
18    53209        8.58
16    46398        7.49
19    39777        6.42
15    38645        6.24
8     34840        5.62
14    34660        5.59
12    32713        5.28
13    31112        5.02
11    30753        4.96
9     29377        4.74
10    27774        4.48
20    27376        4.42
7     26351        4.25
21    19448        3.14
22    16043        2.59
23    13704        2.21
6     12656        2.04
0     11292        1.82
1      9306        1.50
5      8391        1.35
2      7953        1.28
3      6353        1.02
4      6217        1.00

count  percentage
weekday                      
Friday     101785       16.42
Saturday    91496       14.76
Thursday    88693       14.31
Wednesday   88685       14.31
Tuesday     88108       14.22
Monday      82660       13.34
Sunday      78380       12.65

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 619807 entries, 0 to 619806
Data columns (total 46 columns):
 #   Column             Non-Null Count   Dtype         
---  ------             --------------   -----         
 0   num_accident       619807 non-null  int64         
 1   id_vehicle         619807 non-null  int64         
 2   num_vehicle        619807 non-null  object        
 3   category_user      619807 non-null  int64         
 4   injury_severity    619807 non-null  int64         
 5   gender             619807 non-null  int64         
 6   year_of_birth      611268 non-null  float64       
 7   trip_purpose       619807 non-null  int64         
 8   safety_equipment1  619807 non-null  int64         
 9   safety_equipment2  619807 non-null  int64         
 10  safety_equipment3  619807 non-null  int64         
 11  category_vehicle   619807 non-null  int64         
 12  fixed_obs_hit      619807 non-null  int64         
 13  mobile_obs_hit     619807 non-null  int64   

None

In [ ]:
# Save to a single CSV
data.to_csv("01_road_accidents_france_2019_2023_cleaning.csv", sep = ";", index = False)
print("Cleaned file saved as 01_road_accidents_france_2019_2023_cleaning.csv")

Cleaned file saved as 01_road_accidents_france_2019_2023_cleaning.csv


### Handling of `injury_severity` variable

When inspecting the distribution of the variable `injury_severity`, we identified a small subset of records with the value `-1`, corresponding to *"Not specified"* cases. Since this category only represents **0.07 %** (419 out of 619.807) of all records, its influence on the overall analysis is negligible.

To simplify the analysis, these `-1` cases will be included in the **"Non-severe"** category (assigment will happen in collab 02). While this is not perfectly accurate from a semantic standpoint, it allows for a cleaner and more consistent categorization across the dataset without significantly distorting the results.


In [ ]:
# Check distribution of injury severity
checkColumn(data, 'injury_severity', sort=1)



count  percentage
injury_severity                    
 1               260672       42.06
 4               247024       39.85
 3                95248       15.37
 2                16444        2.65
-1                  419        0.07

### Handling of `year_of_birth` (impact on age derivation)

When inspecting the distribution of the variable `year_of_birth`, we identified a subset of records with the value `NaN/-1`, corresponding to *"Not specified"* cases. Since this category represents **1.38 %** (8 539 out of 619.807) of all records, its influence on aggregate results is limited. Note that these records affect the computation of `age` (accident year minus `year_of_birth`) and the corresponding `age_bucket` fields.

To keep the dataset reproducible while avoiding noise in visuals, we retain these `NaN/-1` values in the underlying data but do not display them in age-based charts or KPIs. For derived fields, these cases are treated as **Unknown** (i.e., excluded from `age_bucket` visualizations), which preserves consistency without forcing imputation or distorting the analysis.


In [ ]:
# Check distribution of year_of_birth
checkColumn(data, 'year_of_birth', sort=0)


count  percentage
year_of_birth                   
1900.0           148        0.02
1901.0            43        0.01
1910.0             3        0.00
1911.0             1        0.00
1912.0             1        0.00
...              ...         ...
2020.0          1099        0.18
2021.0           819        0.13
2022.0           509        0.08
2023.0           172        0.03
NaN             8539        1.38

[113 rows x 2 columns]